src/python/pants/backend/java/lint/google_java_format/rules.py

```diff
@@ -0,0 +1,173 @@
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class GoogleJavaFormatFieldSet(FieldSet):
+    required_fields = (JavaSourceField,)
+
+    source: JavaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipGoogleJavaFormatField).value
+
+
+class GoogleJavaFormatRequest(JavaFmtRequest, LintRequest):
+    field_set_type = GoogleJavaFormatFieldSet
+
+
+class GoogleJavaFormatToolLockfileSentinel(JvmToolLockfileSentinel):
+    options_scope = GoogleJavaFormatSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: GoogleJavaFormatRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: Process
+    original_digest: Digest
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_google_java_format(
+    setup_request: SetupRequest,
+    tool: GoogleJavaFormatSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
+) -> Setup:
+    source_files, tool_classpath = await MultiGet(
+        Get(
+            SourceFiles,
+            SourceFilesRequest(field_set.source for field_set in setup_request.request.field_sets),
+        ),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(tool.resolved_lockfile(),),
+            ),
+        ),
+    )
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests([source_files_snapshot.digest, tool_classpath.digest, jdk_setup.digest]),
+    )
+
+    maybe_java16_or_higher_options = []
+    if jdk_setup.jre_major_version >= 16:
+        maybe_java16_or_higher_options = [
+            "--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED",
+            "--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
+            "--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED",
+            "--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
+            "--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED",
+        ]
+
+    maybe_aosp_option = []
+    if tool.aosp:
+        maybe_aosp_option = ["--aosp"]
+
+    args = [
+        *jdk_setup.args(bash, tool_classpath.classpath_entries()),
+        *maybe_java16_or_higher_options,
+        "com.google.googlejavaformat.java.Main",
+        *maybe_aosp_option,
+        "--dry-run" if setup_request.check_only else "--replace",
+        *source_files.files,
+    ]
+
+    process = Process(
+        argv=args,
+        input_digest=input_digest,
+        output_files=source_files_snapshot.files,
```

Comment:
If `check_only`, there is no need to capture the outputs.

Reply:
Context: I was just following the pattern established by the Python linters. For example, see https://github.com/pantsbuild/pants/blob/f9a815028dedb099823b2234a697ce2b373aadd3/src/python/pants/backend/python/lint/black/rules.py#L127

Should we TODO updating all of the linters to not capture outputs?

Reply:
Sure.

Reply:
Going to leave as-is and open an issue regarding changing this pattern everywhere.

Reply:
https://github.com/pantsbuild/pants/issues/13819

Reply:
Oh cool, that's to give better performance by wasting less work? Great idea!